File: Sym_MAS-main/sym_MAS/Simple Discrete System/main1.m
clear all;
close all;

%%%% FOR THE DYNAMICS x(k+1)=x(k)+u_x(k), x(k+1)=x(k)+u_y(k) %%%%

% Data change
n_lim = 5;

% Constraints on state space
bound_x = [1 n_lim];
bound_y = [1 n_lim];
bound = [bound_x; bound_y];

%Constraints on the input space
bound_u = [-2 0; -1 0; 1 0; 2 0; 0 -2; 0 -1; 0 1; 0 2];
n_u = numel(bound_u)/2;

% Space discretization
n_x = n_lim;
n_y = n_lim;

d_x=(bound_x(:,2)-bound_x(:,1))./(n_x); % size of the interval
d_y=(bound_y(:,2)-bound_y(:,1))./(n_y);
d_states = [d_x;d_y];

% Abstraction
disp('')
disp('Step 1: Computation of the abstraction''s transition relation')
disp(' ')

tic

Delta=logical(sparse(n_x*n_y*n_u,n_x*n_y));

% Compute the automata
pb = waitbar(0);
for i1=1:n_x
    for i2=1:n_y
        for h1=1:n_u
            succ_x = i1+bound_u(h1,1);
            succ_y = i2+bound_u(h1,2);
            if((succ_x >= bound(1,1))&&(succ_x <= bound(1,2)) && (succ_y >= bound(2,1))&&(succ_y <= bound(2,2)))
                Delta((i1-1)*n_y*n_u+(i2-1)*n_u+h1, (succ_x-1)*n_y+succ_y) = 1;
        
            end    
        end
    end
end

close(pb);

whos Delta
toc
disp(' ');

% Controller Synthesis
disp('Step 2: Controller synthesis')
disp(' ')
tic
disp('  Fixed point algorithm:')

% Controller Synthesis
Cont=logical(zeros((n_x)*n_y,(n_u))); % create the matrix structure (defined only with 0,1) 
for i=1:(n_x)*n_y % checking admissible modes for any state
    for h=1:(n_u) % for any control mode
        succ=find(Delta((i-1)*n_u+h,:));
        if ~isempty(succ)
          Cont(i,h)=1;
        end
    end
end
ContA=Cont;

iter=0;
%%%% FOR SAFETY SPECIFICATIONS %%%%
% Contp=logical(zeros(n_x*n_y,n_u)); % eliminate transitions reaching blocking states
% while ~all(Cont(:)==Contp(:))
%     iter=iter+1;
%     txt='    Iteration %u\n'; 
%     fprintf(txt,iter)
%     Contp=Cont;

%     % Compute all the successors from controller matrix
%     for i=1:n_x*n_y
%         mode=find(Contp(i,:));
%         if ~isempty(mode)
%             for k=mode
%                 succ=find(Delta((i-1)*n_u+k,:));
%                 if ~isempty(succ)
%                     for ip=succ
%                         if ~any(Contp(ip,:))
%                             Cont(i,k)=0;
%                             break;
%                         end
%                     end
%                 end
%             end
%         end
%     end
% end

t_r1 = [5, 5];

%%%%% FOR REACHABILITY %%%%%
Contp = logical(zeros((n_x)*n_y,(n_u)));
%%% Reachability Specification. Consider a region
reach_states=(t_r1(1)-1)*n_y+t_r1(2);
reach_state_target=reach_states;
while ~all(Cont(:)==Contp(:))
    iter=iter+1;
    txt='               Iteration %u\n'; 
    fprintf(txt,iter)
    Contp=Cont;
    for i=1:(n_x)*n_y
        mode=find(ContA(i,:));
        if ~isempty(mode)
            for k=mode
                succ=find(Delta((i-1)*n_u+k,:));
                if ~isempty(succ)
                    for ip=succ
                        if ~(ismember(ip,reach_states))
                            Cont(i,k)=0;
                        else
                            Cont(i,k)=1;
                            if ~any(i==reach_states)
                                reach_states(end+1) = i;
                            end
                        end
                    end
                end
            end
        end
    end
end

% Converting the Controller into Transition System
controlled_Delta1=logical(sparse(n_x*n_y*n_u,n_x*n_y));
for i=1:(n_x*n_y)
    for h=1:(n_u)
        if(Cont(i,h))
            controlled_Delta1((i-1)*n_u+h,:) = Delta((i-1)*n_u+h,:);
        end
    end
end
toc

disp(' ')
winning_domain=zeros(nnz(Cont),3);
iter = 1;
for i = 1:n_x
    for j = 1:n_y
        for k = 1:n_u
            if(Cont(((i-1)*n_y+j),k))
                winning_domain(iter,:) = [i j k];
                iter=iter+1;
            end
        end
    end
end

save('Controlled_System1','controlled_Delta1');
save('Target_1', 't_r1');

% disp(winning_domain)
init_pos = [3 3];
pos_x=[init_pos(1)];
pos_y=[init_pos(2)];
U_x = find(Cont((init_pos(1)-1)*n_y+init_pos(2),:));
pos=(init_pos(1)-1)*n_y+init_pos(2);

% Simulation
while (reach_state_target~=pos)
%     idx = randi(size(U_x, 2));
    pos=find(Delta((pos-1)*n_u+U_x(1),:));
    if(rem(pos,n_y)==0)
        pos_y_now = n_y;
        pos_y(end+1) = n_y;
    else
        pos_y_now = rem(pos,n_y);
        pos_y(end+1) = rem(pos,n_y);
    end
    pos_x(end+1) = (pos-pos_y_now)/n_y+1;
    U_x = find(Cont(pos,:));
end

plot(pos_x,pos_y)
xlim([1 n_lim])
ylim([1 n_lim])
grid on
grid minor

File: Sym_MAS-main/sym_MAS/Simple Discrete System/main1.py
import numpy as np
import numba as nb
import time

from scipy.io import loadmat,savemat
from scipy.sparse import csr_matrix
from scipy import sparse
from math import pi, sqrt
from sys import exit

# FOR THE DYNAMICS x(k+1)=x(k)+u_x(k), y(k+1)=y(k)+u_y(k)

bound_x = np.array([0, 4])
bound_y = np.array([0, 4])
bound = np.array([bound_x, bound_y])

bound_u = np.array([(-2,0),(-1,0) ,(1,0),(2,0),(0,-2),(0,-1),(0,1),(0,2)])
n_u = len(bound_u)
print("bound_x:", bound_x.shape)
print("bound_y:", bound_y.shape)
print("bound:", bound.shape)
print("bound_u:", bound_u.shape)

# Space discretization
# numbers of intervals
n_x = 5
n_y = 5

# size of the interval
d_x = (bound_x[1] - bound_x[0]) / (n_x)
d_y = (bound_y[1] - bound_y[0]) / (n_y)
d_states = np.array([d_x, d_y])

print("d_x: ", d_x, " d_y: ", d_y)
print("d_states", d_states)

#  Barrier Certificate = d-norm(x_i-x_j)-L
d = 1  # Distance to be maintained
L = 1  # Parameter for abstraction of barrier

# def uni2diff(u):
#     """Convert speed and angular rate to wheel speeds."""
#     v = u[0]
#     omega = u[1]
#     v_L = v - ELL / 2 * omega
#     v_R = v + ELL / 2 * omega
#     return np.array([v_L, v_R]

# Delta = csr_matrix(((n_x*n_y)*(n_x*n_y)*(n_u)*(n_u), (n_x*n_y)*(n_x*n_y)), dtype=bool)
# Delta = Delta2.tolil()
Delta = np.zeros(((n_x*n_y*n_u), (n_x*n_y)))
# @nb.jit(nopython=True, nogil=True)
def compute_succ(Delta):
    for i1 in range(n_x):
        print("Progress:- ", (i1 / n_x) * 100)
        for i2 in range(n_y):                        
            for h1 in range(n_u):
                xp1=i1+bound_u[h1][0]
                yp1=i2+bound_u[h1][1]
                if ((xp1>=bound_x[0]) and (yp1>=bound_y[0]) and (xp1<=bound_x[1]) and (yp1<=bound_y[1])):
                        # associate successors of extrema to extremal intervals
                        row = ((i1)*n_y*n_u + (i2)*n_u+ h1)
                        # print(Delta.shape)
                        # print("Row",row)
                        col = ((xp1)*n_y + (yp1))
                        # print("Column",col)
                        Delta[row,col]=1
    return Delta
# Controller Synthesis
def controller_synthesis(composed_delta):
        Cont = np.zeros((n_x*n_y,n_u),dtype=bool) #create the matrix structure (defined only with 0,1) 
        for i in range(n_x*n_y): #checking admissible modes for any state
            for h in range(n_u): #for any control mode
                succ = np.where(composed_delta[i*n_u+h][:] == 1)
                # print(succ)
                if succ[0].size > 0:
                    Cont[i][h] = 1
        return Cont

# For Safety Specifications
def safety_controller(Controller,composed_delta):
    Contp = np.zeros((n_x*n_y,n_u),dtype=bool) #create the matrix structure (defined only with 0,1) 
    iter=0
    while (not (np.array_equal(Contp, Controller))):
        iter += 1
        Contp = np.copy(Controller)
        for i in range(n_x*n_y):
            print("Progress:- ", (i / (n_x*n_y)) * 100)
            mode = np.where(Contp[i][:] == 1)
            if mode[0].size > 0:
                for k in range(len(mode[0])):
                    val = mode[0][k]
                    succ = np.where(composed_delta[i*n_u + val][:] == 1)
                    if succ[0].size > 0:
                        for ip in range(len(succ[0])):
                            val1 = succ[0][ip]
                            if ~np.any(Contp[val1] == 1):
                                Controller[i][val] = 0
                                break
    return Controller

# For Reachability
def reach_controller(composed_delta,Cont_syn):
    Cont_syn_orig = np.copy(Cont_syn)
    Cont = np.copy(Cont_syn)

    Contp = np.zeros((n_x*n_y,n_u),dtype=bool) #create the matrix structure (defined only with 0,1)
    reach_states = 4*n_y+4

    iter=0
    while (not (np.array_equal(Contp, Cont))):
        iter += 1
        Contp = np.copy(Cont)
        for i in range(n_x*n_y):
            print("Progress:- ", (i / (n_x*n_y)) * 100)
            mode = np.where(Cont_syn_orig[i][:] == 1)
            if mode[0].size > 0:
                for k in range(len(mode[0])):
                    val = mode[0][k]
                    succ = np.where(composed_delta[i*n_u + val][:] == 1)
                    if succ[0].size > 0:    
                        for ip in range(len(succ[0])):
                            val1 = succ[0][ip]
                            if ~(np.isin(i,reach_states)):
                                if ~(np.isin(val1,reach_states)):
                                    Cont[i][val] = 0
                                    break
                                else:
                                    Cont[i][val] = 1
                                    reach_states = np.append(reach_states,i)
        print(np.count_nonzero(Cont))
    return Cont

def controller_transition(Delta,Cont):
    controlled_Delta1 = np.zeros((n_x*n_y*n_u,n_x*n_y),dtype=bool) 
    Delta1 = Delta.copy()
    for i in range(n_x*n_y):
        print("Progress:- ", (i / (n_x*n_y)) * 100)
        for h in range(n_u):
            if(Cont[i][h]):
                controlled_Delta1[i*n_u+h][:] = Delta1[i*n_u+h][:]
            else:
                break
    return controlled_Delta1

# Compute Automata ( reachable sets )
print("Step 1: Computation of the abstraction's transition relation")
start_time = time.time()
control_delta = compute_succ(Delta)
# savemat("path",{'control_delta':control_delta})
print("--- Abstraction Transition total Completion Time %s seconds -- \n" % (time.time() - start_time))

print("Step 2: Computation of Controller")
start_time = time.time()
controller_syn = controller_synthesis(control_delta)
# savemat("path",{'controller_syn':controller_syn})

# print("Step 2: Controller Synthesis for Safety Specification \n")
# start_time = time.time()
# safe_controller = safety_controller(controller_syn,control_delta)
# savemat("path",{'safe_controller':safe_controller})
# print("--- Total Safety Controller Compute Time %s seconds --- \n" % (time.time() - start_time))

print("Step 3: Controller Synthesis for Reachability Specification")
reach_controller = reach_controller(control_delta,controller_syn)
# savemat("path",{'reach_controller':reach_controller})
print("--- Total Reachability Controller Compute Time %s seconds --- \n" % (time.time() - start_time))

print("Step 4: Convert the Controller into a Transition System")
controlled_Delta1 = controller_transition(control_delta,reach_controller)
savemat("D:/Git codes/Sym_MAS/sym_MAS/Simple Discrete System/controlled_Delta1.mat",{'controlled_Delta1':controlled_Delta1})